AppController/djinn.rb

```diff
@@ -1068,28 +1075,33 @@ def upload_app(archived_file, file_suffix, email, secret)
     reservation_id = HelperFunctions.get_random_alphanumeric()
     @app_upload_reservations[reservation_id] = {'status' => 'starting'}
 
-    Djinn.log_debug("Received a request to upload app at #{archived_file}, with suffix #{file_suffix}, with admin user #{email}.")
+    Djinn.log_debug(
+      "Received a request to upload app at #{archived_file}, with suffix " +
+      "#{file_suffix}, with admin user #{email}.")
 
     Thread.new {
+      # If the dashboard is on the same node as the shadow, the archive needs
+      # to be copied to a location that includes the suffix.
       unless archived_file.match(/#{file_suffix}$/)
-        archived_file_old = archived_file
-        archived_file = "#{archived_file_old}.#{file_suffix}"
-        Djinn.log_debug("Renaming #{archived_file_old} to #{archived_file}")
-        File.rename(archived_file_old, archived_file)
+        new_location = [archived_file, file_suffix].join('.')
+        Djinn.log_debug("Copying #{archived_file} to #{new_location}")
+        FileUtils.copy(archived_file, new_location)
+        archived_file = new_location
       end
 
       Djinn.log_debug("Uploading file at location #{archived_file}")
       keyname = @options['keyname']
-      command = "appscale-upload-app --file '#{archived_file}' " +
+      command = "#{UPLOAD_APP_SCRIPT} --file '#{archived_file}' " +
```

Comment:
It's been broken since we stopped modifying the path after installing the tools. It's available on the path (from a shell) automatically after installing with `python setup.py install`, but it's not available from the init script that starts the controller.